intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp

```diff
@@ -195,39 +200,131 @@ std::vector<int> Model::generate_tokens(const std::vector<int>& input_ids) {
-  // Temperature sampling
-  model_sample_top_k(ctx, &candidates_p, top_k, 1);
-  model_sample_tail_free(ctx, &candidates_p, tfs_z, 1);
-  model_sample_typical(ctx, &candidates_p, typical_p, 1);
-  model_sample_top_p(ctx, &candidates_p, top_p, 1);
-  model_sample_temperature(ctx, &candidates_p, temp);
-  int id = model_sample_token(ctx, &candidates_p);
-  return id;
+
+  std::vector<std::pair<double, gpt_vocab::id>> logits_id;
+  logits_id.reserve(n_logits);
+
+  {
+    const float scale = 1.0f / temp;
+    for (int i = 0; i < n_logits; ++i) {
+      // repetition penalty from ctrl paper (https://arxiv.org/abs/1909.05858)
+      // credit https://github.com/facebookresearch/llama/compare/main...shawwn:llama:main
+      if (repeat_last_n > 0 &&
+          std::find(last_n_tokens.end() - repeat_last_n, last_n_tokens.end(), i) != last_n_tokens.end()) {
+        // if score < 0 then repetition penalty has to multiplied to reduce the previous token probability
+        if (plogits[i] < 0.0f) {
+          logits_id.push_back(std::make_pair(plogits[i] * scale * repeat_penalty, i));
+        } else {
+          logits_id.push_back(std::make_pair(plogits[i] * scale / repeat_penalty, i));
+        }
+      } else {
+        logits_id.push_back(std::make_pair(plogits[i] * scale, i));
+      }
+    }
+  }
+
+  // find the top K tokens
+  std::partial_sort(logits_id.begin(), logits_id.begin() + params.top_k, logits_id.end(),
+                    [](const std::pair<double, gpt_vocab::id>& a, const std::pair<double, gpt_vocab::id>& b) {
+                      return a.first > b.first;
+                    });
+
+  logits_id.resize(params.top_k);
+
+  double maxl = -INFINITY;
+  for (const auto& kv : logits_id) {
+    maxl = std::max(maxl, kv.first);
+  }
+
+  // compute probs for the top K tokens
+  std::vector<double> probs;
+  probs.reserve(logits_id.size());
+
+  double sum = 0.0;
+  for (const auto& kv : logits_id) {
+    double p = exp(kv.first - maxl);
+    probs.push_back(p);
+    sum += p;
+  }
+
+  // normalize the probs
+  for (auto& p : probs) {
+    p /= sum;
+  }
+
+  if (params.top_p < 1.0f) {
+    double cumsum = 0.0f;
+    for (int i = 0; i < params.top_k; i++) {
+      cumsum += probs[i];
+      if (cumsum >= params.top_p) {
+        params.top_k = i + 1;
+        probs.resize(params.top_k);
+        logits_id.resize(params.top_k);
+        break;
+      }
+    }
+
+    cumsum = 1.0 / cumsum;
+    for (int i = 0; i < (int)probs.size(); i++) {
+      probs[i] *= cumsum;
+    }
+  }
+
+  //    printf("\n");
+  //    for (int i = 0; i < (int) probs.size(); i++) {
+  //    for (int i = 0; i < 10; i++) {
+  //        printf("%d: '%s' %f\n", i, vocab.id_to_token.at(logits_id[i].second).c_str(), probs[i]);
+  //    }
+
+  std::discrete_distribution<> dist(probs.begin(), probs.end());
+  int idx = dist(rng);
+
+  return logits_id[idx].second;
+}
+
+int Model::post_process(float* logits) {
+  if (params.beam_size > 1) {
```

Comment:
`if (params.beam_size > 1 && !do_sample) `

Reply:
generation mode dispatch reference: https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L849-L887

Reply:
could this work?
```cpp
int Model::post_process(float* logits) {
  if (params.do_sample) {
    return post_sample_top_k_top_p_repeat(logits);
  }

  if (params.beam_size > 1) {
    return post_beam_search(logits);
  }

  return post_greedy_search(logits);
}
```

Reply:
ok

Reply:
thanks, updated